protocol/binary/stream_envelope.go

```diff
@@ -0,0 +1,192 @@
+	return nil
+}
+
+// WriteLegacyEnvelopeBegin writes the start of a non-strict envelope (lacks an envelope version).
+func (sw *StreamWriter) WriteLegacyEnvelopeBegin(eh stream.EnvelopeHeader) error {
+	if err := sw.WriteString(eh.Name); err != nil {
+		return err
+	}
+
+	if err := sw.writeByte(uint8(eh.Type)); err != nil {
+		return err
+	}
+
+	return sw.WriteInt32(eh.SeqID)
+}
+
+// WriteLegacyEnvelopeEnd writes the "end" of a legacy envelope. Since there is
+// no ending to a legacy envelope, this is a no-op.
+func (sw *StreamWriter) WriteLegacyEnvelopeEnd() error {
+	return nil
+}
+
+// ReadEnvelopeBegin reads the start of an Apache Thrift envelope. Thrift supports
+// two kinds of envelopes: strict, and non-strict. See ReadEnveloped method
+// for more information on enveloping.
+func (sw *StreamReader) ReadEnvelopeBegin() (stream.EnvelopeHeader, error) {
+	var eh stream.EnvelopeHeader
+
+	val, err := sw.ReadInt32()
+	if err != nil {
+		return eh, err
+	}
+
+	if val > 0 {
+		if eh, err = sw.readNonStrictEnvelope(val); err != nil {
+			return eh, err
+		}
+	} else {
+		if eh, err = sw.readStrictEnvelope(val); err != nil {
+			return eh, err
+		}
+	}
+
+	seqID, err := sw.ReadInt32()
+	if err != nil {
+		return eh, err
+	}
+
+	eh.SeqID = seqID
+	return eh, nil
+}
+
+// readNonStrictEnvelope reads off a non-strict envelope as described by protocol.EnvelopeAgnosticProtocol.
+func (sw *StreamReader) readNonStrictEnvelope(length int32) (stream.EnvelopeHeader, error) {
+	var eh stream.EnvelopeHeader
+
+	buf := make([]byte, length)
+	for i := int32(0); i < length; i++ {
+		i8, err := sw.ReadInt8()
+		if err != nil {
+			return eh, err
+		}
+		buf[i] = byte(i8)
+	}
+
+	typ, err := sw.ReadInt8()
+	if err != nil {
+		return eh, err
+	}
+
+	eh.Name = string(buf)
+	eh.Type = wire.EnvelopeType(typ)
+	return eh, nil
+}
+
+func (sw *StreamReader) readStrictEnvelope(ver int32) (stream.EnvelopeHeader, error) {
+	var eh stream.EnvelopeHeader
+
+	if v := uint32(ver) & versionMask; v != version1 {
+		return eh, fmt.Errorf("cannot decode envelope of version: %v", v)
+	}
+
+	name, err := sw.ReadString()
+	if err != nil {
+		return eh, err
+	}
+
+	// Casting automatically truncates to the lowest 8 bits.
+	eh.Type = wire.EnvelopeType(ver)
+	eh.Name = name
+	return eh, nil
+}
+
+// ReadEnvelopeEnd reads the "end" of an envelope.  Since there is no real
+// envelope end, this is a no-op.
+func (sw *StreamReader) ReadEnvelopeEnd() error {
+	return nil
+}
+
+type noEnvelopeStreamResponder struct{}
+
+// NoEnvelopeStreamResponder responds to a request without an envelope.
+var NoEnvelopeStreamResponder stream.Responder = &noEnvelopeStreamResponder{}
+
+func (r noEnvelopeStreamResponder) WriteResponseEnvelope(et wire.EnvelopeType, w io.Writer) (stream.Writer, error) {
+	writer := BorrowStreamWriter(w)
+	return writer, nil
+}
+
+// EnvelopeV0StreamResponder responds to requests with a non-strict (un-versioned) envelope.
+type EnvelopeV0StreamResponder struct {
```

Comment:
Do we need new structs for this? Can we:

- Move NoEnvelopeResponder, EnvelopeV0Responder, and EnvelopeV1Responder into the binary package (since they're for the binary protocol)

- Re-export aliases from the protocol package and deprecate them,

    ```go
    // Deprecated: Use binary.EnvelopeV1Responder.
    type EnvelopeV1Responder = binary. EnvelopeV1Responder
    ```

- implement these methods on those types that also implement the non-streaming variants

Reply:
Refactoring the responders into the binary package requires we import "go.uber.org/thriftrw/protocol" and leads to an import cycle. See below for problematic code:

**Import Cycle Code**
```
func (noEnvelopeResponder) EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error {
	return protocol.Binary.Encode(v, w)
}
```

**Import Cycle Error**
```
package go.uber.org/thriftrw
  imports go.uber.org/thriftrw/gen
  imports go.uber.org/thriftrw/internal/plugin
  imports go.uber.org/thriftrw/internal/envelope
  imports go.uber.org/thriftrw/protocol
  imports go.uber.org/thriftrw/protocol/binary
  imports go.uber.org/thriftrw/protocol: import cycle not allowed
```

Let me know what you want to do here.

Reply:
Ah, I see. Because the implementation has to return a `protocol.Responder`. That's unfortunate. Okay, please that suggestion. Exposing the individual responder implementations was probably an oversight.

In the interest of avoiding that issue here, would you consider unexporting all the responder implementations?
This will require moving the ReadRequestEnvelope method from protocol/binary.go into a function inside protocol/binary/ (which the method in protocol/binary.go can just call).

Reply:
Trying to address this in #516. Not tied to that solution so open to feedback/suggestions.
If that seems like too messed up, we can go with the second suggestion in my comment above.

Reply:
Agree with the points you made in #516 - it appears our hands are tied on this one for a "clean" solution. With that said, aliasing+moving the Responders into the binary package and then de-duping the stream enveloper types is still an improvement IMO.

Reply:
Refactored the streaming responder methods onto the existing responder types and moved the responder types+methods into a responder.go file to help organize/clean things up.